framework/doc/content/source/systems/NonlinearSystem.md

```diff
@@ -365,3 +365,64 @@ which there are quite a few nonlinear iterations per timestep, for which the
 cost of an additional Jacobian evaluation during the final residual evaluation
 is amortized. Also, simulations in which material property calculations are very
 expensive may be good candidates for computing the residual and Jacobian together.
+
+## Reusing preconditioners id=reuse_preconditioners
+
+The simple version of GMRES and other iterative methods converge only very
+slowly.  To improve convergence, PETSc and other iterative solver packages
+apply a [preconditioner](https://en.wikipedia.org/wiki/Preconditioner) to the
+system of equations/sparse matrix before applying the iterative solver.
+
+A great number of preconditioners exist, but 
+[multigrid](https://en.wikipedia.org/wiki/Multigrid_method)
+methods are often among the best choices.  
+The [HYPRE](application_development/hypre.md optional=true) package, 
+specifically the 
+BoomerAMG preconditioner, is often a good choice for a preconditioner to
+condition the system of equations resulting from the MOOSE simulation.
+
+A direct factorization of the sparse system of equations is a *very*
+good preconditioner.  An iterative method using the factorized matrix
+as a preconditioner will typically converge to machine precision in a single
+iteration.  However, as noted above, factorizing the sparse system of equations
+for a large simulation is numerically expensive.
+
+One option is to form a preconditioner once and then reuse it to solve
+the linearized system many times.  The preconditioner can be carried over
+through nonlinear iterations and even across time steps.  MOOSE
+allows the user to do this with the `reuse_preconditioner` flag.
+Setting
+
+```
+  reuse_preconditioner = true
+  reuse_preconditioner_max_its = 20
+```
+
+in the `[Executioner]` block will reuse the same preconditioner until
+the number of linear iterations required to solve the linearized system of
+equations exceeds 20.   If the number of linear iterations exceeds 
+`reuse_preconditioner_max_its`
```

Comment:
I missed the boat on the libMesh PR, but I would vote to rename this parameter to `reuse_preconditioner_max_linear_its`. I know that we can set preconditioner evaluation every x number of nonlinear iterations via the PETSc command line option, but maybe in the future we would want a shim option in MOOSE like `reuse_preconditioner_nl_its` (or with a max in there) and then having the distinction between linear-based and nonlinear based would be really nice. Even if we don't ever add any nonlinear based MOOSE Executioner parameters, being more verbose with the the parameter name is hardly a bad thing in my opinion.

Reply:
Yes, I suppose this naming makes more sense.  As we have to change things over there to get AMR to work anyway I'll go ahead and make the change.

Reply:
FYI there's a new sailing boat https://github.com/libMesh/libmesh/pull/3383